Requesting an update

It's been nearly a year since the update, and the project currently fails to compile on Windows 11 using ```1.7.0 stable-x86_64-pc-windows-gnu``` and msys2's CLANG64 package ```mingw-w64-clang-x86_64-rust```.

My three points of interest for an update to jxl-winthumb are to
* Fix errors due to building on newer versions of Rust
* Update libjxl from 0.6.1 to 0.8.1
* Remove the asciidoc dependency (seems unnecessary?)

The errors are as follows:
```
  [ 73%] Built target jxl_enc-obj
  [ 73%] Built target jxl_dec-obj

  --- stderr
  CMake Deprecation Warning at third_party/brotli/CMakeLists.txt:5 (cmake_minimum_required):
    Compatibility with CMake < 2.8.12 will be removed from a future version of
    CMake.

    Update the VERSION argument <min> value or use a ...<max> suffix to tell
    CMake that the project does not need compatibility with older versions.


  -- Compiler is not EMSCRIPTEN
  CMake Deprecation Warning at third_party/sjpeg/CMakeLists.txt:1 (cmake_minimum_required):
    Compatibility with CMake < 2.8.12 will be removed from a future version of
    CMake.

    Update the VERSION argument <min> value or use a ...<max> suffix to tell
    CMake that the project does not need compatibility with older versions.


  CMake Warning at CMakeLists.txt:378 (message):
    sphinx-build not found, skipping rtd documentation


  CMake Warning:
    Manually-specified variables were not used by the project:

      CMAKE_ASM_FLAGS


  SyntaxError: Non-UTF-8 code starting with '\x90' in file C:\Dev\masm\msys64\mingw64\bin\a2x.exe on line 1, but no encoding declared; see https://peps.python.org/pep-0263/ for details
  make[2]: *** [CMakeFiles/manpages.dir/build.make:74: cjxl.1] Error 1
  make[2]: *** Waiting for unfinished jobs....
  SyntaxError: Non-UTF-8 code starting with '\x90' in file C:\Dev\masm\msys64\mingw64\bin\a2x.exe on line 1, but no encoding declared; see https://peps.python.org/pep-0263/ for details
  make[2]: *** [CMakeFiles/manpages.dir/build.make:78: djxl.1] Error 1
  make[1]: *** [CMakeFiles/Makefile2:292: CMakeFiles/manpages.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
  C:/Dev/masm/msys64/opt/cargo/registry/src/index.crates.io-6f17d22bba15001f/libjxl-src-0.7.6/submodules/libjxl/lib/jxl/dec_external_image.cc: In function 'jxl::Status jxl::{anonymous}::ConvertChannelsToExternal(const jxl::ImageF**, size_t, size_t, bool, JxlEndianness, size_t, jxl::ThreadPool*, void*, size_t, const jxl::PixelCallback&, jxl::Orientation)':
  C:/Dev/masm/msys64/opt/cargo/registry/src/index.crates.io-6f17d22bba15001f/libjxl-src-0.7.6/submodules/libjxl/lib/jxl/dec_external_image.cc:290:21: warning: storing the address of local variable 'temp_channels' in '*channels.31_20 + _15' [-Wdangling-pointer=]
    290 |         channels[c] = &(temp_channels[c]);
        |         ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
  C:/Dev/masm/msys64/opt/cargo/registry/src/index.crates.io-6f17d22bba15001f/libjxl-src-0.7.6/submodules/libjxl/lib/jxl/dec_external_image.cc:284:10: note: 'temp_channels' declared here
    284 |   ImageF temp_channels[kConvertMaxChannels];
        |          ^~~~~~~~~~~~~
  C:/Dev/masm/msys64/opt/cargo/registry/src/index.crates.io-6f17d22bba15001f/libjxl-src-0.7.6/submodules/libjxl/lib/jxl/dec_external_image.cc:246:48: note: 'channels' declared here
    246 | Status ConvertChannelsToExternal(const ImageF* channels[], size_t num_channels,
        |                                  ~~~~~~~~~~~~~~^~~~~~~~~~
  make: *** [Makefile:146: all] Error 2
  thread 'main' panicked at '
  command did not execute successfully, got: exit code: 2

  build script failed, must exit now', C:/Dev/masm/msys64/opt/cargo\registry\src\index.crates.io-6f17d22bba15001f\cmake-0.1.48\src\lib.rs:975:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

It's not Rust but with some Python encoding thing per:

```
SyntaxError: Non-UTF-8 code starting with '\x90' in file C:\Dev\masm\msys64\mingw64\bin\a2x.exe on line 1, but no encoding declared; see https://peps.python.org/pep-0263/ for details
  make[2]: *** [CMakeFiles/manpages.dir/build.make:74: cjxl.1] Error 1
  make[2]: *** Waiting for unfinished jobs....
  SyntaxError: Non-UTF-8 code starting with '\x90' in file C:\Dev\masm\msys64\mingw64\bin\a2x.exe on line 1, but no encoding declared; see https://peps.python.org/pep-0263/ for details
  make[2]: *** [CMakeFiles/manpages.dir/build.make:78: djxl.1] Error 1
  make[1]: *** [CMakeFiles/Makefile2:292: CMakeFiles/manpages.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
```

But yeah, whatever that is, JPEGXL_ENABLE_MANPAGES should probably be disabled.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -10,9 +10,7 @@ jobs:
     runs-on: windows-latest
     steps:
       - name: Checkout sources
-        uses: actions/checkout@v2
-        with:
-          submodules: recursive
+        uses: actions/checkout@v3
 
       - name: Install stable toolchain
         uses: actions-rs/toolchain@v1
@@ -20,28 +18,6 @@ jobs:
           profile: minimal
           toolchain: stable
           override: true
-
-      - name: Install Ninja
-        uses: seanmiddleditch/gha-setup-ninja@master
-
-      - name: Cache LLVM and Clang
-        id: cache-llvm
-        uses: actions/cache@v2
-        with:
-          path: ${{ runner.temp }}/llvm
-          key: llvm-14.0
-      - name: Install LLVM and Clang
-        uses: KyleMayes/install-llvm-action@v1
-        with:
-          version: "14.0"
-          directory: ${{ runner.temp }}/llvm
-          cached: ${{ steps.cache-llvm.outputs.cache-hit }}
-
-      # See:
-      # https://github.com/rust-lang/rust-bindgen/issues/1797
-      # https://github.com/KyleMayes/clang-sys/issues/121
-      - name: Set LIBCLANG_PATH
-        run: echo "LIBCLANG_PATH=$((gcm clang).source -replace "clang.exe")" >> $env:GITHUB_ENV
 
       - name: Run cargo build
         uses: actions-rs/cargo@v1
@@ -53,7 +29,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout sources
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
 
       - name: Install stable toolchain
         uses: actions-rs/toolchain@v1

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,8 +12,8 @@ crate-type = ["cdylib"]
 [dependencies]
 simple-logging = "2.0.2"
 log = "0.4.17"
-kagamijxl = "0.3.3"
 winreg = "0.10.1"
+jxl-oxide = "0.3.0"
 
 [dependencies.windows]
 version = "0.37.0"

--- a/README.md
+++ b/README.md
@@ -1,24 +1,10 @@
 # jxl-winthumb
 
-A JPEG XL (*.jxl) thumbnail handler for Windows File Explorer.
-
-Now with WIC decoding support, which means you can use Windows Photo Viewer or any WIC-capable image viewers to view JXL files.
-
-## How to install
-
-0. Install [Microsoft Visual C++ Redistributable for Visual Studio 2022](https://visualstudio.microsoft.com/downloads/#microsoft-visual-c-redistributable-for-visual-studio-2022) if it's not already installed.
-1. Download the latest dll file from the [releases](https://github.com/saschanaz/jxl-winthumb/releases) page.
-2. Open a terminal window as administrator
-3. Move to your download directory
-4. `regsvr32 jxl_winthumb.dll`, or to uninstall, `regsvr32 /u jxl_winthumb.dll`.
+A JPEG XL (*.jxl) WIC decoder to render thumbnails on Windows File Explorer or view images on any WIC-capable image viewers.
 
 ## Build environment
 
-Please read [the requirements](https://github.com/saschanaz/jxl-rs/tree/main/libjxl-src) to build the libjxl dependency, or take a look at [the CI configuration](https://github.com/saschanaz/jxl-winthumb/blob/main/.github/workflows/ci.yml).
-
-### Rust Nightly
-
-Currently Rust Nightly toolchain is required because of [windows-rs dependency](https://github.com/microsoft/windows-rs/issues/1408#issuecomment-1021692345).
+Use the stable Rust toolchain. Current toolchain as of 23th July 2023 is 1.71.0.
 
 ## Helpful resources
 

--- a/src/dll.rs
+++ b/src/dll.rs
@@ -121,7 +121,7 @@ pub unsafe extern "system" fn DllGetClassObject(
     {
         // Set up logging to the project directory.
         simple_logging::log_to_file(
-            &format!("{}\\debug.log", env!("CARGO_MANIFEST_DIR")),
+            format!("{}\\debug.log", env!("CARGO_MANIFEST_DIR")),
             log::LevelFilter::Trace,
         )
         .unwrap();

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -2,7 +2,7 @@
 // https://github.com/microsoft/windows-rs/issues/1506
 #![allow(clippy::not_unsafe_ptr_arg_deref)]
 
-use kagamijxl::{DecodeProgress, Decoder};
+use jxl_oxide::{FrameBuffer, JxlImage, PixelFormat};
 use std::{cell::RefCell, io::BufReader, rc::Rc};
 use windows::core::{implement, Interface, GUID};
 
@@ -22,10 +22,19 @@ mod guid;
 
 mod properties;
 
+type JxlImageFromWinStream = JxlImage<BufReader<WinStream>>;
+
+pub struct DecodedResult {
+    frames: Vec<Rc<FrameBuffer>>,
+    pixel_format: PixelFormat,
+    width: u32,
+    height: u32,
+}
+
 #[implement(Windows::Win32::Graphics::Imaging::IWICBitmapDecoder)]
 #[derive(Default)]
 pub struct JXLWICBitmapDecoder {
-    decoded: RefCell<Option<Rc<RefCell<DecodeProgress>>>>,
+    decoded: RefCell<Option<DecodedResult>>,
 }
 
 impl JXLWICBitmapDecoder {
@@ -50,12 +59,41 @@ impl IWICBitmapDecoder_Impl for JXLWICBitmapDecoder {
         let stream = WinStream::from(pistream.to_owned().unwrap());
         let reader = BufReader::new(stream);
 
-        let decoder = Decoder::new();
-
-        let result = decoder.decode_buffer(reader).map_err(|err| {
+        let mut image = JxlImageFromWinStream::from_reader(reader).map_err(|err| {
             windows::core::Error::new(WINCODEC_ERR_BADIMAGE, format!("{:?}", err).as_str().into())
         })?;
-        self.decoded.replace(Some(Rc::new(RefCell::new(result))));
+        let mut renderer = image.renderer();
+
+        let mut frames: Vec<Rc<FrameBuffer>> = Vec::new();
+
+        let mut load_all = || -> jxl_oxide::Result<usize> {
+            loop {
+                let load_result = renderer.render_next_frame()?;
+                match load_result {
+                    jxl_oxide::RenderResult::NoMoreFrames => {
+                        return Ok(renderer.num_loaded_keyframes())
+                    }
+                    jxl_oxide::RenderResult::NeedMoreData => {
+                        return Err(Box::new(std::io::Error::from(
+                            std::io::ErrorKind::UnexpectedEof,
+                        )))
+                    }
+                    jxl_oxide::RenderResult::Done(render) => {
+                        frames.push(Rc::new(render.image()));
+                    }
+                }
+            }
+        };
+        load_all().map_err(|err| {
+            windows::core::Error::new(WINCODEC_ERR_BADIMAGE, format!("{:?}", err).as_str().into())
+        })?;
+
+        self.decoded.replace(Some(DecodedResult {
+            frames,
+            pixel_format: renderer.pixel_format(),
+            width: renderer.image_header().size.width,
+            height: renderer.image_header().size.height,
+        }));
 
         Ok(())
     }
@@ -78,6 +116,7 @@ impl IWICBitmapDecoder_Impl for JXLWICBitmapDecoder {
 
     fn CopyPalette(&self, _pipalette: &Option<IWICPalette>) -> windows::core::Result<()> {
         log::trace!("JXLWICBitmapDecoder::CopyPalette");
+        // TODO
         WINCODEC_ERR_PALETTEUNAVAILABLE.ok()
     }
 
@@ -93,12 +132,26 @@ impl IWICBitmapDecoder_Impl for JXLWICBitmapDecoder {
 
     fn GetColorContexts(
         &self,
-        _ccount: u32,
-        _ppicolorcontexts: *mut Option<IWICColorContext>,
-        _pcactualcount: *mut u32,
+        ccount: u32,
+        ppicolorcontexts: *mut Option<IWICColorContext>,
+        pcactualcount: *mut u32,
     ) -> windows::core::Result<()> {
-        log::trace!("JXLWICBitmapDecoder::GetColorContexts");
-        WINCODEC_ERR_UNSUPPORTEDOPERATION.ok()
+        log::trace!(
+            "JXLWICBitmapDecoder::GetColorContexts {} {:?} {:?}",
+            ccount,
+            ppicolorcontexts,
+            pcactualcount
+        );
+        // TODO: Proper color context
+        unsafe {
+            if !ppicolorcontexts.is_null() && ccount == 1 {
+                *ppicolorcontexts = Some(JXLWICColorContext {}.into());
+            }
+            if !pcactualcount.is_null() {
+                *pcactualcount = 1;
+            }
+        }
+        Ok(())
     }
 
     fn GetThumbnail(&self) -> windows::core::Result<IWICBitmapSource> {
@@ -107,45 +160,55 @@ impl IWICBitmapDecoder_Impl for JXLWICBitmapDecoder {
     }
 
     fn GetFrameCount(&self) -> windows::core::Result<u32> {
-        if self.decoded.borrow().is_none() {
+        let decoded_ref = self.decoded.borrow();
+        if decoded_ref.is_none() {
             return Err(WINCODEC_ERR_NOTINITIALIZED.ok().unwrap_err());
         }
+        let frame_count = decoded_ref.as_ref().unwrap().frames.len();
 
-        let frame_count = self
-            .decoded
-            .borrow()
-            .as_ref()
-            .unwrap()
-            .borrow()
-            .frames
-            .len();
         log::trace!("JXLWICBitmapDecoder::GetFrameCount: {}", frame_count);
         Ok(frame_count as u32)
     }
 
     fn GetFrame(&self, index: u32) -> windows::core::Result<IWICBitmapFrameDecode> {
-        if self.decoded.borrow().is_none() {
+        let decoded_ref = self.decoded.borrow();
+        if decoded_ref.is_none() {
             return Err(WINCODEC_ERR_NOTINITIALIZED.ok().unwrap_err());
         }
 
-        let basic_info = self.decoded.borrow().as_ref().unwrap().borrow().basic_info;
-        log::trace!("[{}]: {:?}", index, basic_info);
+        let decoded = decoded_ref.as_ref().unwrap();
+        log::trace!("[{}/{}]", index, decoded.frames.len());
 
-        let frame_decode =
-            JXLWICBitmapFrameDecode::new(self.decoded.borrow().to_owned().unwrap(), index as usize);
+        if index >= decoded.frames.len() as u32 {
+            return Err(WINCODEC_ERR_FRAMEMISSING.ok().unwrap_err());
+        }
+
+        let frame_decode = JXLWICBitmapFrameDecode::new(
+            decoded.frames[index as usize].clone(),
+            decoded.pixel_format,
+            decoded.width,
+            decoded.height,
+        );
         Ok(frame_decode.into())
     }
 }
 
 #[implement(Windows::Win32::Graphics::Imaging::IWICBitmapFrameDecode)]
 pub struct JXLWICBitmapFrameDecode {
-    decoded: Rc<RefCell<DecodeProgress>>,
-    index: usize,
+    frame: Rc<FrameBuffer>,
+    pixel_format: PixelFormat,
+    width: u32,
+    height: u32,
 }
 
 impl JXLWICBitmapFrameDecode {
-    pub fn new(decoded: Rc<RefCell<DecodeProgress>>, index: usize) -> Self {
-        Self { decoded, index }
+    pub fn new(frame: Rc<FrameBuffer>, pixel_format: PixelFormat, width: u32, height: u32) -> Self {
+        Self {
+            frame,
+            pixel_format,
+            width,
+            height,
+        }
     }
 }
 
@@ -155,16 +218,30 @@ impl IWICBitmapSource_Impl for JXLWICBitmapFrameDecode {
     fn GetSize(&self, puiwidth: *mut u32, puiheight: *mut u32) -> windows::core::Result<()> {
         log::trace!("JXLWICBitmapFrameDecode::GetSize");
         unsafe {
-            *puiwidth = self.decoded.borrow().basic_info.xsize;
-            *puiheight = self.decoded.borrow().basic_info.ysize;
+            *puiwidth = self.width;
+            *puiheight = self.height;
         }
         Ok(())
     }
 
     fn GetPixelFormat(&self) -> windows::core::Result<GUID> {
         log::trace!("JXLWICBitmapFrameDecode::GetPixelFormat");
-        // TODO: Support HDR
-        Ok(GUID_WICPixelFormat32bppRGBA)
+        // TODO: Support all formats
+        match self.pixel_format {
+            PixelFormat::Gray => Ok(GUID_WICPixelFormat32bppGrayFloat),
+            PixelFormat::Graya => Err(windows::core::Error::new(
+                WINCODEC_ERR_UNSUPPORTEDPIXELFORMAT,
+                "Gray alpha image is currently not supported".into(),
+            )),
+            PixelFormat::Rgb => Ok(GUID_WICPixelFormat96bppRGBFloat),
+            PixelFormat::Rgba => Ok(GUID_WICPixelFormat128bppRGBAFloat),
+            jxl_oxide::PixelFormat::Cmyk | jxl_oxide::PixelFormat::Cmyka => {
+                Err(windows::core::Error::new(
+                    WINCODEC_ERR_BADIMAGE,
+                    "Cmyk is currently not supported".into(),
+                ))
+            }
+        }
     }
 
     fn GetResolution(&self, pdpix: *mut f64, pdpiy: *mut f64) -> windows::core::Result<()> {
@@ -198,17 +275,19 @@ impl IWICBitmapSource_Impl for JXLWICBitmapFrameDecode {
         let prc = unsafe { prc.as_ref().unwrap() };
         log::trace!("JXLWICBitmapFrameDecode::CopyPixels::WICRect {:?}", prc);
 
-        let basic_info = &self.decoded.borrow().basic_info;
-        let data = &self.decoded.borrow().frames[self.index].data;
+        let channels = self.frame.channels();
 
         for y in prc.Y..(prc.Y + prc.Height) {
-            let src_offset = basic_info.xsize as i32 * 4 * y;
+            let src_offset = self.width as i32 * channels as i32 * y;
             let dst_offset = prc.Width * 4 * (y - prc.Y);
             unsafe {
                 std::ptr::copy_nonoverlapping(
-                    data.as_ptr().offset((src_offset + prc.X) as isize),
-                    pbbuffer.offset(dst_offset as isize),
-                    (prc.Width as usize) * 4,
+                    self.frame
+                        .buf()
+                        .as_ptr()
+                        .offset((src_offset + prc.X) as isize),
+                    (pbbuffer as *mut f32).offset(dst_offset as isize),
+                    (prc.Width as usize) * channels,
                 );
             }
         }
@@ -225,13 +304,23 @@ impl IWICBitmapFrameDecode_Impl for JXLWICBitmapFrameDecode {
 
     fn GetColorContexts(
         &self,
-        _ccount: u32,
-        _ppicolorcontexts: *mut Option<IWICColorContext>,
+        ccount: u32,
+        ppicolorcontexts: *mut Option<IWICColorContext>,
         pcactualcount: *mut u32,
     ) -> windows::core::Result<()> {
-        log::trace!("JXLWICBitmapFrameDecode::GetColorContexts");
+        log::trace!(
+            "JXLWICBitmapFrameDecode::GetColorContexts {} {:?} {:?}",
+            ccount,
+            ppicolorcontexts,
+            pcactualcount
+        );
         unsafe {
-            *pcactualcount = 0;
+            if !ppicolorcontexts.is_null() && ccount == 1 {
+                *ppicolorcontexts = Some(JXLWICColorContext {}.into());
+            }
+            if !pcactualcount.is_null() {
+                *pcactualcount = 1;
+            }
         }
         Ok(())
     }
@@ -239,5 +328,47 @@ impl IWICBitmapFrameDecode_Impl for JXLWICBitmapFrameDecode {
     fn GetThumbnail(&self) -> windows::core::Result<IWICBitmapSource> {
         log::trace!("JXLWICBitmapFrameDecode::GetThumbnail");
         Err(WINCODEC_ERR_CODECNOTHUMBNAIL.ok().unwrap_err())
+    }
+}
+
+#[implement(Windows::Win32::Graphics::Imaging::IWICColorContext)]
+pub struct JXLWICColorContext {}
+
+impl IWICColorContext_Impl for JXLWICColorContext {
+    fn InitializeFromFilename(
+        &self,
+        _wzfilename: &::windows::core::PCWSTR,
+    ) -> ::windows::core::Result<()> {
+        WINCODEC_ERR_UNSUPPORTEDOPERATION.ok()
+    }
+
+    fn InitializeFromMemory(
+        &self,
+        _pbbuffer: *const u8,
+        _cbbuffersize: u32,
+    ) -> ::windows::core::Result<()> {
+        WINCODEC_ERR_UNSUPPORTEDOPERATION.ok()
+    }
+
+    fn InitializeFromExifColorSpace(&self, _value: u32) -> ::windows::core::Result<()> {
+        WINCODEC_ERR_UNSUPPORTEDOPERATION.ok()
+    }
+
+    fn GetType(&self) -> ::windows::core::Result<WICColorContextType> {
+        Ok(WICColorContextExifColorSpace)
+    }
+
+    fn GetProfileBytes(
+        &self,
+        _cbbuffer: u32,
+        _pbbuffer: *mut u8,
+        _pcbactual: *mut u32,
+    ) -> ::windows::core::Result<()> {
+        // TODO: Implement this for proper ICC profile support
+        WINCODEC_ERR_UNSUPPORTEDOPERATION.ok()
+    }
+
+    fn GetExifColorSpace(&self) -> ::windows::core::Result<u32> {
+        Ok(1) // sRGB
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,6 +1,9 @@
 #![crate_type = "dylib"]
 // https://github.com/microsoft/windows-rs/issues/1506
 #![allow(clippy::not_unsafe_ptr_arg_deref)]
+// TODO: Update windows-rs
+#![allow(unused_must_use)]
+#![allow(non_snake_case)]
 
 use jxl_oxide::{FrameBuffer, JxlImage, PixelFormat};
 use std::{cell::RefCell, io::BufReader, rc::Rc};

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -91,11 +91,19 @@ impl IWICBitmapDecoder_Impl for JXLWICBitmapDecoder {
             windows::core::Error::new(WINCODEC_ERR_BADIMAGE, format!("{:?}", err).as_str().into())
         })?;
 
+        let (width, height, _left, _top) = renderer.image_header().metadata.apply_orientation(
+            renderer.image_header().size.width,
+            renderer.image_header().size.height,
+            0,
+            0,
+            false,
+        );
+
         self.decoded.replace(Some(DecodedResult {
             frames,
             pixel_format: renderer.pixel_format(),
-            width: renderer.image_header().size.width,
-            height: renderer.image_header().size.height,
+            width,
+            height,
         }));
 
         Ok(())

--- a/src/properties.rs
+++ b/src/properties.rs
@@ -55,7 +55,7 @@ impl IInitializeWithStream_Impl for JXLPropertyStore {
         unsafe {
             PSCreateMemoryPropertyStore(
                 &IPropertyStoreCache::IID,
-                std::mem::transmute(&self.props),
+                &self.props as *const _ as *mut *mut std::ffi::c_void,
             )?
         };
 

--- a/src/properties.rs
+++ b/src/properties.rs
@@ -46,11 +46,18 @@ impl IInitializeWithStream_Impl for JXLPropertyStore {
         let stream = WinStream::from(pstream.to_owned().unwrap());
         let reader = BufReader::new(stream);
 
-        let decoder = JxlImage::from_reader(reader).map_err(|err| {
+        let mut image = JxlImage::from_reader(reader).map_err(|err| {
             windows::core::Error::new(WINCODEC_ERR_BADIMAGE, format!("{:?}", err).as_str().into())
         })?;
+        let renderer = image.renderer();
 
-        let size = &decoder.image_header().size;
+        let (width, height, _left, _top) = renderer.image_header().metadata.apply_orientation(
+            renderer.image_header().size.width,
+            renderer.image_header().size.height,
+            0,
+            0,
+            false,
+        );
 
         unsafe {
             PSCreateMemoryPropertyStore(
@@ -67,14 +74,14 @@ impl IInitializeWithStream_Impl for JXLPropertyStore {
         let PSGUID_IMAGESUMMARYINFORMATION =
             GUID::from_u128(0x6444048F_4C8B_11D1_8B70_080036B11A03);
 
-        let variant = unsafe { InitPropVariantFromUInt32Vector(&[size.width])? };
+        let variant = unsafe { InitPropVariantFromUInt32Vector(&[width])? };
         let propkey = PROPERTYKEY {
             fmtid: PSGUID_IMAGESUMMARYINFORMATION,
             pid: 3,
         };
         unsafe { props.SetValueAndState(&propkey, &variant, PSC_READONLY)? };
 
-        let variant = unsafe { InitPropVariantFromUInt32Vector(&[size.height])? };
+        let variant = unsafe { InitPropVariantFromUInt32Vector(&[height])? };
         let propkey = PROPERTYKEY {
             fmtid: PSGUID_IMAGESUMMARYINFORMATION,
             pid: 4,
@@ -92,9 +99,8 @@ impl IInitializeWithStream_Impl for JXLPropertyStore {
             let pcwstr = pcwstr.into_param().abi();
             InitPropVariantFromStringVector(&[PWSTR(pcwstr.0 as *mut _)])
         }
-        let variant = unsafe {
-            InitPropVariantFromStringVectorWrapped(format!("{} x {}", size.width, size.height))?
-        };
+        let variant =
+            unsafe { InitPropVariantFromStringVectorWrapped(format!("{} x {}", width, height))? };
         let propkey = PROPERTYKEY {
             fmtid: PSGUID_IMAGESUMMARYINFORMATION,
             pid: 13,

--- a/src/properties.rs
+++ b/src/properties.rs
@@ -14,7 +14,7 @@ use windows::Win32::{
 };
 
 use crate::winstream::WinStream;
-use kagamijxl::Decoder;
+use jxl_oxide::JxlImage;
 
 #[implement(
     Windows::Win32::UI::Shell::PropertiesSystem::IInitializeWithStream,
@@ -46,11 +46,11 @@ impl IInitializeWithStream_Impl for JXLPropertyStore {
         let stream = WinStream::from(pstream.to_owned().unwrap());
         let reader = BufReader::new(stream);
 
-        let mut decoder = Decoder::new();
-        decoder.no_full_image = true;
-        let result = decoder.decode_buffer(reader).map_err(|err| {
+        let decoder = JxlImage::from_reader(reader).map_err(|err| {
             windows::core::Error::new(WINCODEC_ERR_BADIMAGE, format!("{:?}", err).as_str().into())
         })?;
+
+        let size = &decoder.image_header().size;
 
         unsafe {
             PSCreateMemoryPropertyStore(
@@ -67,14 +67,14 @@ impl IInitializeWithStream_Impl for JXLPropertyStore {
         let PSGUID_IMAGESUMMARYINFORMATION =
             GUID::from_u128(0x6444048F_4C8B_11D1_8B70_080036B11A03);
 
-        let variant = unsafe { InitPropVariantFromUInt32Vector(&[result.basic_info.xsize])? };
+        let variant = unsafe { InitPropVariantFromUInt32Vector(&[size.width])? };
         let propkey = PROPERTYKEY {
             fmtid: PSGUID_IMAGESUMMARYINFORMATION,
             pid: 3,
         };
         unsafe { props.SetValueAndState(&propkey, &variant, PSC_READONLY)? };
 
-        let variant = unsafe { InitPropVariantFromUInt32Vector(&[result.basic_info.ysize])? };
+        let variant = unsafe { InitPropVariantFromUInt32Vector(&[size.height])? };
         let propkey = PROPERTYKEY {
             fmtid: PSGUID_IMAGESUMMARYINFORMATION,
             pid: 4,
@@ -93,10 +93,7 @@ impl IInitializeWithStream_Impl for JXLPropertyStore {
             InitPropVariantFromStringVector(&[PWSTR(pcwstr.0 as *mut _)])
         }
         let variant = unsafe {
-            InitPropVariantFromStringVectorWrapped(format!(
-                "{} x {}",
-                result.basic_info.xsize, result.basic_info.ysize
-            ))?
+            InitPropVariantFromStringVectorWrapped(format!("{} x {}", size.width, size.height))?
         };
         let propkey = PROPERTYKEY {
             fmtid: PSGUID_IMAGESUMMARYINFORMATION,

--- a/src/registry.rs
+++ b/src/registry.rs
@@ -21,7 +21,7 @@ const PERCEIVED_TYPE_VALUE: &str = "image";
 fn register_clsid_base(module_path: &str, clsid: &windows::core::GUID) -> std::io::Result<RegKey> {
     let hkcr = RegKey::predef(HKEY_CLASSES_ROOT);
     let clsid_key = hkcr.open_subkey("CLSID")?;
-    let (key, _) = clsid_key.create_subkey(&guid_to_string(clsid))?;
+    let (key, _) = clsid_key.create_subkey(guid_to_string(clsid))?;
     key.set_value("", &"jxl-winthumb")?;
 
     let (inproc, _) = key.create_subkey("InProcServer32")?;
@@ -167,7 +167,7 @@ fn register_provider() -> std::io::Result<()> {
     // https://docs.microsoft.com/en-us/windows/win32/wic/-wic-integrationregentries#integration-with-the-windows-thumbnail-cache
     let (system_shell_ex, _) = system_ext_key.create_subkey("ShellEx")?;
     system_shell_ex
-        .create_subkey(&guid_to_string(
+        .create_subkey(guid_to_string(
             &windows::Win32::UI::Shell::IThumbnailProvider::IID,
         ))?
         .0

--- a/src/registry/property_handler.rs
+++ b/src/registry/property_handler.rs
@@ -29,7 +29,7 @@ pub fn unregister_property_handler() -> std::io::Result<()> {
 
     let clsid_key = hkcr.open_subkey("CLSID")?;
     clsid_key
-        .delete_subkey_all(&guid_to_string(&JXLPropertyStore::CLSID))
+        .delete_subkey_all(guid_to_string(&JXLPropertyStore::CLSID))
         .ok();
 
     let hklm = RegKey::predef(HKEY_LOCAL_MACHINE);

